File: Tri_rapide/tri_rapide.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int* read_file(char* filename, int* taille){
    FILE* fichier = NULL;
    fichier = fopen(filename, "r");
    int* tab = malloc(*taille * sizeof(int));
    *taille = 0;
    while(fscanf(fichier, "%d",&tab[*taille]) == 1){
        (*taille) ++;
    }
    fclose(fichier);
    return tab;
}

int partition(int* tab, int p, int r) {
    int pivot = tab[r-1];
    int i = p;
    int* buf = malloc(1 * sizeof(int));
    buf[0] = 0;
    for(int j=p; j < r-1;j++)
    {
        if(tab[j] <= pivot)
        {
            buf[0] = tab[i];
            tab[i] = tab[j];
            tab[j] = buf[0];
            i++;
        }
    }
    buf[0] = tab[i];
    tab[i] = tab[r-1];
    tab[r-1] = buf[0];

    free(buf);
    return i;
}

int* sous_tri_rapide(int* tab, int p, int r) {
    if(p<r-1)
    {
        int q = partition(tab,p,r);
        sous_tri_rapide(tab, p, q);
        sous_tri_rapide(tab, q+1, r);
    }
    return tab;
}


int* tri_rapide(char* filename, int taille){
    int* tab = read_file(filename,&taille);
    tab = sous_tri_rapide(tab,0,taille);
    for(int i=0; i<taille; i++)
    {
        printf("%d\n", tab[i]);
    }
    return tab;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <size>\n", argv[0]);
        return 1;
    }

    int size = atoi(argv[1]);
    int* tab = tri_rapide("./tab.txt",size);
    free(tab);

    return 0;
}

File: Tri_rapide/tri_rapide.py
import argparse
import sys
def read_file(filename):
    valeurs = []
    try:
        with open(filename, 'r') as fichier:
            contenu = fichier.read().strip()  # Lire le contenu du fichier et supprimer les espaces en excès
            valeurs = contenu.split()  # Diviser le contenu en valeurs individuelles en utilisant les espaces comme séparateurs
            valeurs = [float(val) for val in valeurs]  # Convertir les valeurs en flottants
    except FileNotFoundError:
        print("Le fichier spécifié n'a pas été trouvé.")
    except Exception as e:
        print("Une erreur est survenue :", e)
    return valeurs

def partition(tab, p, r):
    pivot = tab[r-1]
    i = p
    for j in range(p, r-1):
        if tab[j] < pivot:
            temp = tab[i]
            tab[i] = tab[j]
            tab[j] = temp
            i+=1
    temp = tab[i]
    tab[i] = tab[r-1]
    tab[r-1] = temp
    return i

def sousTriRapide(tab, p, r):
    if p < r-1:
        q = partition(tab, p, r)
        sousTriRapide(tab, p, q)
        sousTriRapide(tab, q+1, r)


def triRapide(filename, size):
    tab = read_file(filename)
    print(tab)
    sousTriRapide(tab, 0, size)
    print(tab)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Effectue un tri rapide sur un fichier spécifié.")
    parser.add_argument("filename", type=str, help="Chemin vers le fichier à trier.")
    parser.add_argument("size", type=int, help="Taille du tableau à trier.")
    args = parser.parse_args()

    # Définissez la limite de profondeur de récursion
    sys.setrecursionlimit(args.size)  # Vous pouvez ajuster la limite selon vos besoins

    triRapide(args.filename, args.size)